Onboarding Wizard: Add Welcome page

yscik opened the pull request:
Fixes #3037

### Changes proposed in this Pull Request

* Add a welcome page to the Onboarding wizard
* Add a REST API endpoint to `Sensei_Onboarding`
 * This routes to a method based on a `page` parameter in the URL, and request method.
 * Current URL: `/sensei/v1/onboarding/<page>`
* Get and set usage tracking data in the appropiate REST-callable methods.
* Move `Sensei_Admin::create_pages` so we don't have to instantiate `Sensei_Admin` in REST calls. 

### Testing instructions

* Delete Courses and My courses pages. 
* Navigate to Sensei LMS > Onboarding in wp-admin
* Observe the page
* Click *"Continue"*
* *"Yes, count me in"* checkbox in the modal that pops up should be checked if the user previously opted in to usage tracking. 
* Change the checkbox value, click *"Continue"*
* Usage tracking value should change (Can be confired in Sensei LMS > Settings, General tab, last item)
* Courses and My Courses pages should be created.

### Tasks

- [x] Add page, modal
- [x] Set up custom colors
- [x] Use REST API to get & set tracking state
- [x] Create base Sensei pages if they don't exists 
- [x] Cancel onboarding with 'Not right now'
- [ ] Configure i18n, pot generation | Moving this to a dedicated card
- [x] Advance to the next step after submit | Can be added after #3085 is merged

### Deprecated Code
- `Sensei_Admin::create_page` - Replaced by `Sensei_Onboarding_Pages::create_page`
- `Sensei_Admin::create_pages` - Replaced by `Sensei_Onboarding_Pages::create_pages`

### Screenshot

![Onboarding welcome](https://user-images.githubusercontent.com/176949/80499353-2ac1b580-896d-11ea-9ea3-e0a522125917.gif)

renatho commented on assets/onboarding/welcome.jsx:
```diff
@@ -0,0 +1,61 @@
+import { Card, H, Link } from '@woocommerce/components';
```
I was using one component (and support components) per folder structure - in this case, I'd create `welcome/index.jsx`. I like it because if we have some specific style file, test file, stories, or other things related, we can add it together. And I think it's a good structure to grow organized. What do you think? If you have a different opinion, we can discuss and I can refactor where I used this structure.

yscik replied:
Good approach, I moved them in bb6d287.

renatho commented on includes/class-sensei-admin.php:
```diff
@@ -228,6 +228,8 @@ function install_pages_output() {
 	/**
 	 * create_page function.
 	 *
+	 * @deprecated since 4.0.0 use Sensei_onboarding\Pages::create_page
```
Should it be `3.1.0`? Another thing is that we need to remember to add it to the issue to remove the deprecation code ;)

yscik replied:
Yup, fixed in b935750.

yscik commented on assets/onboarding/style.scss:
```diff
@@ -26,13 +28,79 @@
 				margin: 0;
 			}
+
+      &__step.is-complete, &__step.is-active {
```
Moved and renamed it to `wp-colors.scss` to make it a bit clearer that it's for the external components.

yscik commented on assets/onboarding/usage-modal.jsx:
```diff
@@ -0,0 +1,68 @@
+ * @param {Function} props.onContinue Callback for user pressing the continue button.
+ * @param {Function} props.onClose Callback for closing the modal.
+ * @param {boolean} props.isSubmitting Indicate loading state.
+ * @class
```
I looked into this but couldn't find an ESLint rule for it. But we should, looks more consistent. ed4c036

renatho replied:
Hum. I tried to look it up too, but I didn't find any rule too. :(

yscik commented on assets/onboarding/usage-modal.jsx:
```diff
@@ -0,0 +1,68 @@
+ * @param {boolean} props.isSubmitting Indicate loading state.
+ * @class
+ */
+export function UsageModal( { tracking, onContinue, onClose, isSubmitting } ) {
```
Updated in e00c216.

yscik commented on assets/onboarding/use-onboarding-api.jsx:
```diff
@@ -0,0 +1,48 @@
+	const [ isBusy, setBusy ] = useState( false );
+	let path = `/sensei/v1/onboarding/${ step }`;
+	useEffect( () => {
+		fetchData();
```
I added a `useCallback` in 9e5561a. This implementation might not be very long-lived as the recommended way to do all this would be through a `wp-data` store, instead of directly using `apiFetch`, so we should probably move to that at some point.

yscik commented on assets/onboarding/use-onboarding-api.jsx:
```diff
@@ -0,0 +1,48 @@
+import { useEffect, useState } from '@wordpress/element';
```
Yup, deb93e6.

yscik commented:
I added some PHP unit tests for the REST API endpoints. 
Button styles and positions should also be fixed now.

donnapep commented on assets/onboarding/welcome/usage-modal.jsx:
```diff
@@ -0,0 +1,72 @@
+	const trackingMessage = interpolateComponents( {
+		mixedString: __(
+			'Get improved features and faster fixes by sharing non-sensitive data via {{link}}usage tracking{{/link}} ' +
+				'that shows us how WooCommerce is used. No personal data is tracked or stored.',
```
Gasp! WooCommerce! 😆

yscik replied:
Oops! How did that happen? 30aac2e

donnapep commented:
When closing the modal the button has the focus. Can that be removed?

![Screen Shot 2020-05-05 at 2 01 29 PM](https://user-images.githubusercontent.com/1190420/81099490-0029ac80-8ed9-11ea-948e-0287dd710e99.jpg)

When I set _Course Archive Page_ and _My Courses Page_ to _Select a page_ (without deleting the actual pages), and then clicked _Continue_ on the modal, the following PHP notice was logged:
```
PHP Notice:  Undefined index: usage_tracking in /Users/donnapep/Local Sites/sensei/app/public/wp-content/plugins/sensei-lms/includes/admin/class-sensei-onboarding.php on line 183
```

Also, when the new Courses page is created, it's blank on the front-end. When I re-save the permalinks the page starts showing. Clicking the old _Install Sensei LMS Pages_ button works fine.

renatho commented on includes/class-sensei.php:
```diff
@@ -397,6 +397,9 @@ public function initialize_global_objects() {
 		$this->enrolment_scheduler->init();
 
+		// Onboarding Wizard.
+		$this->onboarding = new Sensei_Onboarding();
```
Oh, now that I saw that you put the onboarding out of the `if`. Is it only for tests purpose? If yes, maybe should we do it differently? Else, I'll refactor the test in the other PR to use it.

yscik replied:
This is needed for the REST API, because those requests are not invoking the WP-Admin screen.

renatho replied:
Oh, perfect! Tomorrow I'll refactor the tests in the other PR to use like this =)

yscik commented on assets/onboarding/welcome/index.jsx:
```diff
@@ -0,0 +1,63 @@
+				</p>
+				<p>
+					{ __(
+						'This setup wizard will help you get started creating online courses more quickly. It is optional and should only take a few minutes.',
```
Updated in d02d3ec

yscik commented on assets/onboarding/welcome/index.jsx:
```diff
@@ -0,0 +1,63 @@
+import { __ } from '@wordpress/i18n';
+import { Button } from '@wordpress/components';
+import { useState } from '@wordpress/element';
+import { UsageModal } from './usage-modal.jsx';
```
We can live without it probably :) Cleaned it up in 02ea776.

yscik commented:
> Also, when the new Courses page is created, it's blank on the front-end. When I re-save the permalinks the page starts showing. 

Added the command to flush rewrite rules after updating the settings - right now in Sensei this will run on the next request for a Wp-admin page, so when the user leaves the Onboarding wizard after the Ready step, in a normal flow, as REST API request don't trigger this. 

> When closing the modal the button has the focus. Can that be removed?

This is an intentional accessibility feature of the modal implementation, restoring focus to what was focused before it opened. Cannot be removed easily. 

> PHP Notice:  Undefined index: usage_tracking

Fixed in 6567625.

yscik commented on includes/admin/class-sensei-onboarding-pages.php:
```diff
@@ -0,0 +1,73 @@
+ * Create Sensei pages during onboarding.
+ *
+ * @package Sensei\Onboarding
+ * @since   1.3.0
```
Oops! df58f55

donnapep commented on includes/class-sensei-admin.php:
```diff
@@ -228,6 +228,8 @@ function install_pages_output() {
 	/**
 	 * create_page function.
 	 *
+	 * @deprecated since 3.1.0 use Sensei_onboarding\Pages::create_page
```
Looks like the class is called `Sensei_Onboarding_Pages`.

yscik commented:
Rebased to master.

alexsanford commented on assets/onboarding/welcome/index.jsx:
```diff
@@ -0,0 +1,63 @@
+ * Welcome step for Onboarding Wizard.
+ */
+export const Welcome = () => {
+	const [ usageModalActive, toggleUsageModal ] = useState( false );
```
Why do we need blank lines between these variable declarations?

alexsanford commented on assets/onboarding/use-onboarding-api.js:
```diff
@@ -0,0 +1,51 @@
+			data: formData,
+		} );
+		setBusy( false );
+		await fetchData();
```
Could we architect this in a way that doesn't require a second call here? Maybe the POST itself could return the new data?

alexsanford commented on includes/admin/class-sensei-onboarding.php:
```diff
@@ -155,4 +163,85 @@ public function add_onboarding_help_tab( $screen ) {
+	/**
+	 * Register REST API route.
+	 */
+	public function register_rest_api() {
```
We should look at the REST API architecture. I think it would be best to use the pattern of of REST API Controller classes to create the endpoints. Open to other options, of course. See WCPC for an example of how we architected it there.

alexsanford commented on includes/admin/class-sensei-onboarding.php:
```diff
@@ -155,4 +163,85 @@ public function add_onboarding_help_tab( $screen ) {
+	public function register_rest_api() {
+
+		register_rest_route(
+			'sensei/v1',
```
I also mentioned this elsewhere, but I think we should use a namespace like `sensei-internal` for this.

alexsanford commented on includes/class-sensei-admin.php:
```diff
@@ -237,48 +239,27 @@ function install_pages_output() {
-		);
 
-		$page_id = wp_insert_post( $page_data );
+		_deprecated_function( __METHOD__, '3.1.0', 'Sensei_Onboarding_Pages::create_page' );
```
Was this deprecation added to the appropriate issue for removal later?

alexsanford commented on includes/class-sensei-usage-tracking.php:
```diff
@@ -65,7 +65,7 @@ public function get_tracking_enabled() {
 	}
 
-	protected function set_tracking_enabled( $enable ) {
+	public function set_tracking_enabled( $enable ) {
```
I wonder if we should mark the `@access` for this?

alexsanford commented on includes/class-sensei-assets.php:
```diff
@@ -160,4 +160,23 @@ public function asset_config( $filename, $dependencies = [], $args = null ) {
+	 *
+	 * @param string[] $rest_routes REST routes to preload.
+	 */
+	public function preload_data( $rest_routes ) {
```
Where is this method used?

alexsanford commented on tests/unit-tests/admin/test-class-sensei-onboarding-api.php:
```diff
@@ -0,0 +1,149 @@
+	 *
+	 * @covers Sensei_Onboarding::handle_api_request
+	 */
+	public function testOnlyAdminUserCanAccessOnboardingAPI() {
```
This should be broken into two tests, because it is testing two things (Teacher cannot access, admin can access).

alexsanford commented on tests/unit-tests/admin/test-class-sensei-onboarding-api.php:
```diff
@@ -0,0 +1,149 @@
+	 */
+	public function testSubmitWelcomeCreatesSenseiPages() {
+
+		$this->request( 'POST', 'welcome', [ 'usage_tracking' => false ] );
```
Should we ensure that the pages do not exist before this call, to be sure?
